### ISSUE TITLE
Convert MPLRegressor with multiple outputs

### ISSUE THREAD

**Issue by baumfalk** [jpmml/jpmml-sklearn#91]
I  have created an MLPRegressor that functions as an auto-encoder:
```
clf = MLPRegressor(hidden_layer_sizes = [100,50,100], verbose=True, activation='logistic')
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('clf', clf) ])
pipeline.fit(X, X)
```
Here, X is an array of 10000 images with dimension 100x100 (which are stacked on top of each other)
```
$ print(X.shape)
(10000, 10000)
```

The output of a prediction is also an image (with the pixels stack on top of each other)

```
$ print(y)
(1, 10000)
```
However, I can't export this model to a PMML model:
```
from sklearn2pmml import PMMLPipeline, make_pmml_pipeline, sklearn2pmml
PMML_pipeline = make_pmml_pipeline(pipeline)
sklearn2pmml(PMML_pipeline, 'pipeline.pmml', debug = True)
```
gives

```
python: 3.6.6
sklearn: 0.20.0
sklearn.externals.joblib: 0.12.5
pandas: 0.23.4
sklearn_pandas: 1.7.0
sklearn2pmml: 0.39.0
java: 1.8.0_191

WARNING: Attribute 'sklearn2pmml.pipeline.PMMLPipeline.target_fields' is not set. Assuming y as the name of the target field
Nov 16, 2018 1:29:34 PM sklearn2pmml.pipeline.PMMLPipeline initFeatures
WARNING: Attribute 'sklearn2pmml.pipeline.PMMLPipeline.active_fields' is not set. Assuming [x1, ...., x10000] as the names of active fields

Nov 16, 2018 1:29:34 PM org.jpmml.sklearn.Main run
SEVERE: Failed to convert
java.lang.IllegalArgumentException
	at org.jpmml.converter.neural_network.NeuralNetworkUtil.createRegressionNeuralOutputs(NeuralNetworkUtil.java:160)
	at sklearn.neural_network.MultilayerPerceptronUtil.encodeNeuralNetwork(MultilayerPerceptronUtil.java:144)
	at sklearn.neural_network.MLPRegressor.encodeModel(MLPRegressor.java:49)
	at sklearn.neural_network.MLPRegressor.encodeModel(MLPRegressor.java:29)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:219)
	at org.jpmml.sklearn.Main.run(Main.java:145)
	at org.jpmml.sklearn.Main.main(Main.java:94)

Exception in thread "main" java.lang.IllegalArgumentException
	at org.jpmml.converter.neural_network.NeuralNetworkUtil.createRegressionNeuralOutputs(NeuralNetworkUtil.java:160)
	at sklearn.neural_network.MultilayerPerceptronUtil.encodeNeuralNetwork(MultilayerPerceptronUtil.java:144)
	at sklearn.neural_network.MLPRegressor.encodeModel(MLPRegressor.java:49)
	at sklearn.neural_network.MLPRegressor.encodeModel(MLPRegressor.java:29)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:219)
	at org.jpmml.sklearn.Main.run(Main.java:145)
	at org.jpmml.sklearn.Main.main(Main.java:94)
```

I have also tried explicitely setting the target fields:

```
from sklearn2pmml import PMMLPipeline, make_pmml_pipeline, sklearn2pmml
PMML_pipeline = make_pmml_pipeline(pipeline,
                                   target_fields=list(range(1,X.shape[0]+1)))
sklearn2pmml(PMML_pipeline, 'pipeline.pmml', debug = True)
```

But that gave the following error:
```
SEVERE: Failed to convert
java.lang.IllegalArgumentException: Expected 1 element(s), got 10000 element(s)
	at org.jpmml.sklearn.ClassDictUtil.checkSize(ClassDictUtil.java:63)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:115)
	at org.jpmml.sklearn.Main.run(Main.java:145)
	at org.jpmml.sklearn.Main.main(Main.java:94)

Exception in thread "main" java.lang.IllegalArgumentException: Expected 1 element(s), got 10000 element(s)
	at org.jpmml.sklearn.ClassDictUtil.checkSize(ClassDictUtil.java:63)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:115)
	at org.jpmml.sklearn.Main.run(Main.java:145)
	at org.jpmml.sklearn.Main.main(Main.java:94)
```

What am I doing wrong? Thank you in advance for taking the time to help me!

**Comment by vruusmann** [comment]
> What am I doing wrong?

You're doing everything right from the Scikit-Learn perspective.

This `java.lang.IllegalArgumentException: Expected 1 element(s), got X element(s)` simply means that multi-target models are not yet supported ("Expected 1 element(s)" really means "Expected 1 target field(s)" in this context).

This issue is related to https://github.com/jpmml/sklearn2pmml/issues/54

The current model schema descriptor class `org.jpmml.converter.Schema` needs to be extended for multi-target case. I haven't found a satisfactory API solution to this problem yet - should this class be extended (eg. `class MultiSchema extends Schema`), or be kept as-is and extended at the label class `org.jpmml.converter.Label` level instead (eg. `class MultiContinuousLabel extends ContinuousLabel`). 

As a general note - you appear to be training rather wide neural networks (100x50x100) using unstructured data (images). The PMML data format is really not meant for this, although technically it should work. You might consider the [ONNX](https://onnx.ai) data format instead.

**Comment by baumfalk** [comment]
Thank you for your swift response! I will find another way to do it (ONNX might help with that).

**Comment by christophe-rannou** [comment]
I am trying to solve the multi output problem for regression problems.

In particular I started with trying to enforce multiple outputs for a Ridge regression with two outputs.

My strategy is to implement the `org.jpmml.converter.MultiContinuousLabel` class and to replace the `sklearn.linear_model.LinearRegressor` by a `sklearn.linear_model.MultiLinearRegressor` object which encodes the model as a Segment for each output as the target with a `org.dmg.pmml.regression.RegressionModel` for each segment. For the Segmentation I chose SELECT_ALL as the `org.dmg.pmml.mining.Segmentation.MultipleModelMethod`.

During evaluation I end up in the selectAll method of the class `org.jpmml.evaluator.mining.MiningModelEvaluator`. In this method I do no get why each `org.jpmml.evaluator.mining.SegmentResult` is expected to have the exact same keySet or even the same length. Could you give me some details ?

In addition if you have any feedback regarding my approach I would be grateful. You can find the WIP implementation of the jpmml-converter [here](https://github.com/christophe-rannou/jpmml-converter/tree/feature/multi-output-regressor-support) and of the jpmml-sklearn [here](https://github.com/christophe-rannou/jpmml-sklearn/tree/feature/multi-output-regressor-support).

Thanks
